### PR TITLE
fix(focus-trap): add mutation observer

### PR DESCRIPTION
**Pull request by sjohnson-sage**
Fixes: #4255

### Proposed behaviour
Add `MutationObserver` to `FocusTrap` to reset it when something in it's child tree changes.

### Current behaviour
The `FocusTrap` does not reset if it's children changes.

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [ ] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
Sandbox in issue rebuilt with this code in the codesandbox comment below

### CODE REVIEW

File: src/__internal__/focus-trap/focus-trap.component.js [hunk 1]
```diff
@@ -53,7 +54,26 @@ const FocusTrap = ({
...
       }
     }
-  }, [children, hasNewInputs, wrapperRef]);
+  }, [hasNewInputs, wrapperRef]);
+
+  useEffect(() => {
+    const observer = new MutationObserver(function () {
+      updateFocusableElements();
+    });
+
```
Comment (mmaciorowski-sage) [inline src/__internal__/focus-trap/focus-trap.component.js:63]: I think `const observer = new MutationObserver(updateFocusableElements);` should work

Comment (sjohnson-sage) [reply src/__internal__/focus-trap/focus-trap.component.js:63]: Good point 😅

File: src/__internal__/focus-trap/focus-trap.spec.js [hunk 2]
```diff
@@ -428,4 +429,59 @@ describe("FocusTrap", () => {
...
+  describe("when content in the children tree changes", () => {
+    it("should trigger the MutationObserver", () => {
+      const mutationObserverMock = jest.fn(function MutationObserver(callback) {
+        this.observe = jest.fn();
+        this.disconnect = jest.fn();
+        this.trigger = () => {
+          callback();
+        };
+      });
+      global.MutationObserver = mutationObserverMock;
```
Comment (edleeks87) [inline src/__internal__/focus-trap/focus-trap.spec.js:442]: Do we not need to reset `global.MutationObserver` after the test has run?

Comment (sjohnson-sage) [reply src/__internal__/focus-trap/focus-trap.spec.js:442]: We do indeed

**Review by edleeks87 — APPROVED** [review]

**Review by mmaciorowski-sage — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Pull request merged**